auth/postgres/groups.go

```diff
@@ -13,471 +13,493 @@ import (
+			case errDuplicate:
+				return auth.Group{}, errors.Wrap(auth.ErrGroupConflict, err)
+			}
+		}
+		return auth.Group{}, errors.Wrap(auth.ErrUpdateGroup, errors.New(pqErr.Message))
 	}
 
-	return g, nil
+	defer row.Close()
+	row.Next()
+	dbu = dbGroup{}
+	if err := row.StructScan(&dbu); err != nil {
+		return g, errors.Wrap(auth.ErrUpdateGroup, err)
+	}
+
+	return toGroup(dbu)
 }
 
 func (gr groupRepository) Delete(ctx context.Context, groupID string) error {
 	qd := `DELETE FROM groups WHERE id = :id`
-	group := groups.Group{
+	group := auth.Group{
 		ID: groupID,
 	}
 	dbg, err := gr.toDBGroup(group)
 	if err != nil {
-		return errors.Wrap(errUpdateDB, err)
+		return errors.Wrap(auth.ErrUpdateGroup, err)
 	}
 
 	res, err := gr.db.NamedExecContext(ctx, qd, dbg)
 	if err != nil {
-		return errors.Wrap(errDeleteGroupDB, err)
+		pqErr, ok := err.(*pq.Error)
+		if ok {
+			switch pqErr.Code.Name() {
+			case errInvalid, errTruncation:
+				return errors.Wrap(auth.ErrMalformedEntity, err)
+			case errFK:
+				switch pqErr.Constraint {
+				case groupIDFkeyy:
+					return errors.Wrap(auth.ErrGroupNotEmpty, err)
+				}
+				return errors.Wrap(auth.ErrGroupConflict, err)
+			}
+		}
+		return errors.Wrap(auth.ErrUpdateGroup, errors.New(pqErr.Message))
 	}
 
 	cnt, err := res.RowsAffected()
 	if err != nil {
-		return errors.Wrap(errDeleteGroupDB, err)
+		return errors.Wrap(auth.ErrDeleteGroup, err)
 	}
 
 	if cnt != 1 {
-		return errors.Wrap(groups.ErrDeleteGroup, err)
+		return errors.Wrap(auth.ErrDeleteGroup, err)
 	}
 	return nil
 }
 
-func (gr groupRepository) RetrieveByID(ctx context.Context, id string) (groups.Group, error) {
+func (gr groupRepository) RetrieveByID(ctx context.Context, id string) (auth.Group, error) {
 	dbu := dbGroup{
 		ID: id,
 	}
-
-	q := `SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM groups WHERE id = $1`
+	q := `SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level, created_at, updated_at FROM groups WHERE id = $1`
 	if err := gr.db.QueryRowxContext(ctx, q, id).StructScan(&dbu); err != nil {
 		if err == sql.ErrNoRows {
-			return groups.Group{}, errors.Wrap(groups.ErrNotFound, err)
+			return auth.Group{}, errors.Wrap(auth.ErrGroupNotFound, err)
 
 		}
-		return groups.Group{}, errors.Wrap(errRetrieveDB, err)
+		return auth.Group{}, errors.Wrap(auth.ErrSelectEntity, err)
 	}
-
 	return toGroup(dbu)
 }
 
-func (gr groupRepository) RetrieveAll(ctx context.Context, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
-	_, mq, err := getGroupsMetadataQuery("groups", gm)
+func (gr groupRepository) RetrieveAll(ctx context.Context, pm auth.PageMetadata) (auth.GroupPage, error) {
+	_, mq, err := getGroupsMetadataQuery("groups", pm.Metadata)
 	if err != nil {
-		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+		return auth.GroupPage{}, errors.Wrap(auth.ErrFailedToRetrieveAll, err)
 	}
+	cq := "SELECT COUNT(*) FROM groups"
```

Comment:
A simple optimization: You can move the count query below `rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)` so that we don't instantiate it in case that first query returns an error.